### PR TITLE
#160609550 Return descriptive validation errors at registration

### PR DESCRIPTION
**Pull request by samuelbwambale**
### What does this PR do?

Add descriptive descriptive validation error messages on user registration

#### Description of Task to be completed?

Currently, when a user is signing up, they get default error messages which are generated by the of DRF which are not descriptive enough for the Authors Haven app

This task ensures that users get specific error responses on validating the user inputs. The PR covers both invalid and valid valid user parameters 

#### How should this be manually tested?

`POST /api/users`

Registration details:

```source-json
{
  "user":{
    "username": "Jacob",
    "email": "jake@jake.jake",
    "password": "jakejake"
  }
}
```

Registration with invalid email:

```source-json
{
  "user":{
    "username": "Jacob",
    "email": "jakejake.jake",
    "password": "jakejake"
  }
}
```

#### Any background context you want to provide?

N/A

#### Screenshots

screen shot 2018-09-28 at 11 45 36

#### What are the relevant pivotal tracker stories?

[#160609550](https://www.pivotaltracker.com/story/show/160609550)

### CODE REVIEW

File: authors/apps/authentication/views.py [hunk 1]
```diff
@@ -18,11 +20,26 @@ class RegistrationAPIView(APIView):
...
+            )
+        elif len(user['username']) < 4:
+            raise serializers.ValidationError(
+                'The username should be atleast 4 characters'
+            )
+        elif len(user['password']) < 8:
+            raise serializers.ValidationError(
+                'The password should be atleast 8 characters'
+            )
+        if bool(re.match('(?=.*[0-9])(?=.*[a-zA-Z])', user['password'])) is False:
```
Comment (isheebo) [inline authors/apps/authentication/views.py]:
```python
if re.match(...., ....) is None:
   raise ...
```

Comment (samuelbwambale) [reply authors/apps/authentication/views.py]: This has been fixed and pushed

**Review by kakaemma — APPROVED** [review]
Nice work @samuelbwambale

File: authors/settings.py [hunk 2]
```diff
@@ -84,9 +84,9 @@
 # Database
 # https://docs.djangoproject.com/en/1.11/ref/settings/#databases
 
-DATABASE_URL = config('DATABASE_URL')
-DATABASES = {}
-DATABASES['default'] = dj_database_url.config(conn_max_age=600, ssl_require=True)
+DATABASES = {
```
Comment (lym) [inline authors/settings.py]: @samuelbwambale How is this change relevant to the issue that your PR seeks to address?

Comment (samuelbwambale) [reply authors/settings.py]: @lym I was not able to register a user locally with the previous database settings. I changed the settings to test that the feature works as I expected. The app functions properly both locally and on Heroku

Comment (lym) [reply authors/settings.py]: @samuelbwambale Did the app not work for anyone till this point?

Comment (samuelbwambale) [reply authors/settings.py]: The issue started after I rebased from develop this morning. @MuhanguziDavid experienced the same while working on JWT Authentication.
The app was starting normally but on registering a user, I was getting an error `raise ImproperlyConfigured("settings.DATABASES is improperly configured. "`

Comment (lym) [reply authors/settings.py]: The proper way to do this would've been to create two separate PRs; the first one fixing the database configuration issue and the second implementing your story. I cannot merge this PR as it stands.

Comment (samuelbwambale) [reply authors/settings.py]: Thanks @lym.  [#160897544](https://www.pivotaltracker.com/story/show/160897544) has been raised on PT board to fix the database configuration issue

Comment (samuelbwambale) [reply authors/settings.py]: @lym Issue fixed

File: authors/settings.py [hunk 3]
```diff
@@ -86,7 +86,7 @@
 
 DATABASES = {
    'default': dj_database_url.config(default=config('DATABASE_URL')
-)}
+)}
```
Comment (lym) [inline authors/settings.py]: Get rid of this trailing space.

**Comment by samuelbwambale** [comment]
@lym the trailing spaces have been deleted

**Review by mendozabree — COMMENTED** [review]
Create a separate method for password validation.

**Comment by samuelbwambale** [comment]
@mendozabree I have created a validators.py file to handle validation of email, password and username

File: authors/apps/authentication/views.py [hunk 4]
```diff
@@ -1,13 +1,12 @@
...
 from rest_framework.generics import RetrieveUpdateAPIView
 from rest_framework.permissions import AllowAny, IsAuthenticated
 from rest_framework.response import Response
 from rest_framework.views import APIView
 
 from .renderers import UserJSONRenderer
-from .serializers import (
-    LoginSerializer, RegistrationSerializer, UserSerializer
-)
+from .serializers import (LoginSerializer, RegistrationSerializer,
```
Comment (lym) [inline authors/apps/authentication/views.py]: Please rollback this change.

Comment (samuelbwambale) [reply authors/apps/authentication/views.py]: @lym changes done

**Comment by lym** [comment]
@samuelbwambale Please look into the failing build.

**Comment by samuelbwambale** [comment]
@lym build fixed. It failed because of failed validations in tests.

**Comment by lym** [comment]
👍
